pkg/diff/renderer_test.go

```diff
@@ -204,3 +204,141 @@ func TestRenderPrimitiveValue(t *testing.T) {
 		})
 	}
 }
+
+func TestRenderByCommand(t *testing.T) {
+	t.Parallel()
+
+	diffCommand := "diff"
+
+	type SubStruct struct {
+		FieldSub string
+	}
+	type TestStruct struct {
+		StringV     string
+		StringP     *string
+		IntV        int
+		BoolV       bool
+		TaggedV     int `json:"taggedv_tag"`
+		StringSlice []string
+		Map         map[string]string
+		SubStruct   SubStruct
+	}
+
+	testcases := []struct {
+		name         string
+		command      string
+		old          TestStruct
+		new          TestStruct
+		expectedDiff string
+		expectErr    bool
+	}{
+		{
+			name:         "invalid command",
+			command:      "invalid-command",
+			old:          TestStruct{},
+			new:          TestStruct{},
+			expectedDiff: "",
+			expectErr:    true,
+		},
+		{
+			name:    "no diff",
+			command: diffCommand,
+			old: TestStruct{
+				StringV:     "value1",
+				StringP:     nil,
+				IntV:        1,
+				BoolV:       true,
+				TaggedV:     1,
+				StringSlice: []string{"a", "b"},
+				Map: map[string]string{
+					"key1": "value1",
+				},
+				SubStruct: SubStruct{
+					FieldSub: "valueSub",
+				},
+			},
+			new: TestStruct{
+				StringV:     "value1",
+				StringP:     nil,
+				IntV:        1,
+				BoolV:       true,
+				TaggedV:     1,
+				StringSlice: []string{"a", "b"},
+				Map: map[string]string{
+					"key1": "value1",
+				},
+				SubStruct: SubStruct{
+					FieldSub: "valueSub",
+				},
+			},
+			expectedDiff: "",
+			expectErr:    false,
+		},
+		{
+			name:    "has diff",
+			command: diffCommand,
+			old: TestStruct{
+				StringV:     "value1",
+				StringP:     &[]string{"a"}[0],
+				IntV:        1,
+				BoolV:       true,
+				TaggedV:     1,
+				StringSlice: []string{"a", "b"},
+				Map: map[string]string{
+					"key1": "value1",
+				},
+				SubStruct: SubStruct{
+					FieldSub: "valueSub",
+				},
+			},
+			new: TestStruct{
+				StringV:     "value2",
+				StringP:     &[]string{"b"}[0],
+				IntV:        2,
+				BoolV:       false,
+				TaggedV:     10,
+				StringSlice: []string{"a", "b", "c"},
+				Map: map[string]string{
+					"key1": "valueXXX",
+				},
+				SubStruct: SubStruct{
+					FieldSub: "xxx",
+				},
+			},
+			expectedDiff: `@@ -1,12 +1,13 @@
+-BoolV: true
+-IntV: 1
++BoolV: false
++IntV: 2
+ Map:
+-  key1: value1
+-StringP: a
++  key1: valueXXX
++StringP: b
+ StringSlice:
+ - a
+ - b
+-StringV: value1
++- c
++StringV: value2
+ SubStruct:
+-  FieldSub: valueSub
+-taggedv_tag: 1
++  FieldSub: xxx
++taggedv_tag: 10`,
+			expectErr: false,
+		},
+	}
+
+	for _, tc := range testcases {
+		t.Run(tc.name, func(t *testing.T) {
+			actual, err := RenderByCommand(tc.command, tc.old, tc.new)
+			if tc.expectErr {
+				assert.Error(t, err)
+			} else {
+				assert.NoError(t, err)
+			}
```

Comment:
[nit]
```suggestion
			assert.Equal(t, tc.expectErr, err != nil)
```

Reply:
@ffjlabo 
Thanks!!! I fixed it: [64efe0a](https://github.com/pipe-cd/pipecd/pull/4881/commits/64efe0a6bfacb595c8638d9c4ffd4e9860f77ffc)